Add execution_role_arn parameter

mjmayer opened the pull request:
##### SUMMARY
Fargate support requires containers to have the executionRoleArn defined. This PR adds that parameter.
Fixes #41848
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ecs_taskdefinition

##### ANSIBLE VERSION
```
ansible 2.7.0.dev0 (devel 46dbb3e1e8) last updated 2018/06/22 10:32:41 (GMT -700)
  config file = None
  configured module search path = [u'/home/mjmayer/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/mjmayer/ansible/lib/ansible
  executable location = /home/mjmayer/ansible/bin/ansible
  python version = 2.7.5 (default, Apr 11 2018, 17:41:36) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28.0.1)]

```

mjmayer committed:
Add execution_role_arn parameter

mjmayer committed:
Change ecs_taskdefinition to use AnsibleAWSmodule

Botocore version checking is becomming more common. Changing the ecs_taskdefinition
to use AnsibleAWSmodule allows more easily for this.

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -319,6 +326,10 @@ def main():
             module.fail_json(msg='botocore needs to be version 1.8.4 or higher to use launch_type')
 
+    if module.params['task_role_arn']:
+        if not module.botocore_at_least('1.10.44'):
```
Thanks for adding this! Do you mind doing the same for launch_type above by replacing `if not task_mgr.ecs_api_supports_requirescompatibilities():` with `if not module.botocore_at_least('1.8.4'):`?

mjmayer committed:
Change launch type check to use botocore_at_least function

mjmayer committed:
Remove execution_role_arn param from params dict

willthames commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -225,6 +230,7 @@ def register_task(self, family, task_role_arn, network_mode, container_definitio
         params = dict(
             family=family,
             taskRoleArn=task_role_arn,
+            executionRoleArn=execution_role_arn,
```
executionRoleArn must not be set if it's not required, otherwise anyone relying on this module will need to upgrade to a very recent botocore.

willthames commented:
It's worth noting that this fails the existing test suite. 

```
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1529926266.6214693-137085483742155/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_e53r2y2w/ansible_module_ecs_taskdefinition.py", line 488, in <module>
    main()
  File "/tmp/ansible_e53r2y2w/ansible_module_ecs_taskdefinition.py", line 456, in main
    module.params['memory'])
  File "/tmp/ansible_e53r2y2w/ansible_module_ecs_taskdefinition.py", line 246, in register_task
    response = self.ecs.register_task_definition(**params)
  File "/usr/lib/python3.6/site-packages/botocore/client.py", line 312, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/lib/python3.6/site-packages/botocore/client.py", line 579, in _make_api_call
    api_params, operation_model, context=request_context)
  File "/usr/lib/python3.6/site-packages/botocore/client.py", line 634, in _convert_to_request_dict
    api_params, operation_model)
  File "/usr/lib/python3.6/site-packages/botocore/validate.py", line 291, in serialize_to_request
    raise ParamValidationError(report=report.generate_report())
botocore.exceptions.ParamValidationError: Parameter validation failed:
Unknown parameter in input: "executionRoleArn", must be one of: family, taskRoleArn, networkMode, containerDefinitions, volumes, placementConstraints

fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_e53r2y2w/ansible_module_ecs_taskdefinition.py\", line 488, in <module>\n    main()\n  File \"/tmp/ansible_e53r2y2w/ansible_module_ecs_taskdefinition.py\", line 456, in main\n    module.params['memory'])\n  File \"/tmp/ansible_e53r2y2w/ansible_module_ecs_taskdefinition.py\", line 246, in register_task\n    response = self.ecs.register_task_definition(**params)\n  File \"/usr/lib/python3.6/site-packages/botocore/client.py\", line 312, in _api_call\n    return self._make_api_call(operation_name, kwargs)\n  File \"/usr/lib/python3.6/site-packages/botocore/client.py\", line 579, in _make_api_call\n    api_params, operation_model, context=request_context)\n  File \"/usr/lib/python3.6/site-packages/botocore/client.py\", line 634, in _convert_to_request_dict\n    api_params, operation_model)\n  File \"/usr/lib/python3.6/site-packages/botocore/validate.py\", line 291, in serialize_to_request\n    raise ParamValidationError(report=report.generate_report())\nbotocore.exceptions.ParamValidationError: Parameter validation failed:\nUnknown parameter in input: \"executionRoleArn\", must be one of: family, taskRoleArn, networkMode, containerDefinitions, volumes, placementConstraints\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 1
}
```

willthames commented:
Could we have a test for this change?

mjmayer committed:
Change check to use parameter

mjmayer committed:
Fix typo

mjmayer committed:
Add test for old botocore version

mjmayer committed:
Add test for execution role parameter

willthames commented on test/integration/targets/ecs_cluster/playbooks/roles/ecs_cluster/tasks/main.yml:
```diff
@@ -541,6 +541,23 @@
+        - AmazonEC2ContainerServiceRole
+        <<: *aws_connection_info
+
+    - name: gather facts for the ecsTaskExecutionRole
```
Should be able to access this from `register`ing the results of the previous task

mjmayer replied:
I registered the output from iam_role task and reran the tests. The tests passed.

mjmayer committed:
Remove iam_role_facts task

Task was unecessary. The same information could be gathered by registering
the iam_role task.

willthames commented:
Merged, thanks @mjmayer
